shared/src/balancer_event_handler.rs

```diff
@@ -3,195 +3,183 @@ use crate::{
     event_handling::{BlockNumber, EventHandler, EventIndex, EventStoring},
     impl_event_retrieving,
     maintenance::Maintaining,
+    Web3,
 };
 use anyhow::{anyhow, Context, Result};
 use contracts::{
-    balancer_v2_vault::{
-        self,
-        event_data::{
-            PoolRegistered as ContractPoolRegistered, TokensRegistered as ContractTokensRegistered,
-        },
-        Event as ContractEvent,
+    balancer_v2_factory::{
+        self, event_data::PoolCreated as ContractPoolCreated, Event as ContractEvent,
     },
-    BalancerV2Vault,
+    BalancerV2Vault, BalancerV2WeightedPool, BalancerV2WeightedPoolFactory,
 };
+use derivative::Derivative;
 use ethcontract::common::DeploymentInformation;
-use ethcontract::{dyns::DynWeb3, Event as EthContractEvent, EventMetadata, H160, H256};
+use ethcontract::{
+    dyns::DynWeb3, Bytes, Event as EthContractEvent, EventMetadata, H160, H256, U256,
+};
+use mockall::*;
 use std::{
     collections::{HashMap, HashSet},
     fmt::Debug,
     ops::RangeInclusive,
 };
 use tokio::sync::Mutex;
 
-#[derive(Clone, Debug)]
-pub enum BalancerEvent {
-    PoolRegistered(PoolRegistered),
-    TokensRegistered(TokensRegistered),
-}
-
-#[derive(Debug, Clone, PartialEq)]
-pub struct PoolRegistered {
-    pub pool_id: H256,
+#[derive(Copy, Debug, Default, Clone, Eq, PartialEq, Hash)]
+pub struct PoolCreated {
     pub pool_address: H160,
-    pub specialization: PoolSpecialization,
-}
-
-#[derive(Debug, Clone, PartialEq)]
-pub struct TokensRegistered {
-    pub pool_id: H256,
-    pub tokens: Vec<H160>,
 }
 
 #[derive(Clone, Debug, Eq, PartialEq, Hash)]
 pub struct WeightedPool {
     pool_id: H256,
     pool_address: H160,
     tokens: Vec<H160>,
-    specialization: PoolSpecialization,
+    normalized_weights: Vec<U256>,
     block_created: u64,
 }
 
-#[derive(Clone, Debug, Default, PartialEq)]
+#[derive(Clone, Copy, Debug, Default, Eq, PartialEq, Hash)]
 pub struct WeightedPoolBuilder {
```

Comment:
Is this still needed now that we have a single event?

Reply:
Well, we do still need to make the web3 requests for additional data, and this would allow us to parallelize/batch those queries. What do you think?

Reply:
But what advantage does this have over just using `(PoolCreated, u64)`?

Reply:
Ok sure, we may not need a `WeightedPoolBuilder` struct anymore, but if we don't keep it in pending_pools, where are we going to keep it? Or should we just perform try_upgrade on the list of tuples immediately? I see what you mean here.

Reply:
My point is that we can create the `WeighedPool` directly from the `PoolCreated` event.

Reply:
> Or should we just perform try_upgrade on the list of tuples immediately? I see what you mean here.

:+1: